lib/theme_check/language_server/variable_lookup_finder.rb

```diff
@@ -1,51 +1,48 @@
 # frozen_string_literal: true
 
+require 'ostruct'
+
 module ThemeCheck
   module LanguageServer
     module VariableLookupFinder
+      include Constants
       extend self
 
-      UNCLOSED_SQUARE_BRACKET = /\[[^\]]*\Z/
-      ENDS_IN_BRACKET_POSITION_THAT_CANT_BE_COMPLETED = %r{
-        (
-          # quotes not preceded by a [
-          (?<!\[)['"]|
-          # closing ]
-          \]|
-          # opening [
-          \[
-        )$
-      }x
-
-      VARIABLE_LOOKUP_CHARACTERS = /[a-z0-9_.'"\]\[]/i
-      VARIABLE_LOOKUP = /#{VARIABLE_LOOKUP_CHARACTERS}+/o
-      SYMBOLS_PRECEDING_POTENTIAL_LOOKUPS = %r{
-        (?:
-          \s(?:
-            if|elsif|unless|and|or|#{Liquid::Condition.operators.keys.join("|")}
-            |echo
-            |case|when
-            |cycle
-            |in
-          )
-          |[:,=]
-        )
-        \s+
-      }omix
-      ENDS_WITH_BLANK_POTENTIAL_LOOKUP = /#{SYMBOLS_PRECEDING_POTENTIAL_LOOKUPS}$/oimx
-      ENDS_WITH_POTENTIAL_LOOKUP = /#{SYMBOLS_PRECEDING_POTENTIAL_LOOKUPS}#{VARIABLE_LOOKUP}$/oimx
+      PotentialLookup = Struct.new(:name, :lookups)
+
+      def lookup(token)
+        content = token.content
+        cursor = token.cursor
 
-      def lookup(content, cursor)
         return if cursor_is_on_bracket_position_that_cant_be_completed(content, cursor)
-        potential_lookup = lookup_liquid_variable(content, cursor) || lookup_liquid_tag(content, cursor)
+        variable_lookup = lookup_liquid_variable(content, cursor) || lookup_liquid_tag(content, cursor)
 
         # And we only return it if it's parsed by Liquid as VariableLookup
-        return unless potential_lookup.is_a?(Liquid::VariableLookup)
-        potential_lookup
+        return unless variable_lookup.is_a?(Liquid::VariableLookup)
+
+        potential_lookup(variable_lookup, token)
       end
 
       private
 
+      def potential_lookup(variable, token)
+        return variable if token.buffer.nil? || token.buffer.empty?
+
+        buffer = token.buffer[0...token.absolute_cursor]
+        lookups = variable.lookups
+        assignments = find_assignments(buffer)
```

Comment:
Thanks for raising that, @charlespwd!

I had something similar to the `AssignmentsManager` running on `textDocument/didChange` in mind, still when I started implementing it I noticed some details that weren't looking so good:

- I. `AssignmentsManager` would visit Liquid templates on every file change (regardless that's used for code completion or not)
- II. Also, it would require some index management to handle multiple files
- III. Also, it would be almost required to have some kind of memoization mechanism to avoid re-visiting tokens (as they would happen a bit frequently) \*

With those points in mind, I've moved the `AssignmentsFinder` to the `textDocument/completion` phase, and then we execute it only when it's required.

Do you have any thoughts about that?

<br>

\--

<br>

> Some of the concerns I'd have with the other approach is handling things like for x in list and if else statements. e.g.
> ```liquid
> {% liquid
>  if type == "product"
>     assign x = product
>  elsif some_condition
>     # kind of convoluted example, but this part of the tree 
>     # technically doesn't have a type for `x`, and I'd be worried
>     # about something that looks at all the code until now to break.
>     assign y = x.value
>  endif
> %}
> ```

I believe we may introduced some level of scope support on the `AssignmentsFinder` to handle iterations, so it would depend on the new ✨ `CompletionContext` instead of `buffer`.

With that enhancement, it would get the type of `p` when we can infer the collection type, and also the `last_product` type:

```liquid
{% for p in collections.first.products %}
  {% assign last_product = p %}
  {{ p.title }}
{% endfor %}

{{ last_product.█ }}
```

Still, I don't think we can should cover this kind of scenario (https://github.com/Shopify/theme-check/issues/655), as we'd need SFR/runtime-context to have confidence about the type of `target` (which depends on `some_condition`):

```liquid
{%- liquid
  if some_condition
    assign target = cart
  else
    assign target = product
  endif
%}
{{ target.█ }}
```

Do you have any extra thoughts about that?

Thanks again for raising those points, @charlespwd! :)